fedimint-recurringd/src/db.rs

```diff
@@ -0,0 +1,185 @@
+    pub fn random() -> FederationDbPrefix {
+        FederationDbPrefix(thread_rng().gen())
+    }
+
+    fn prepend(&self, byte: u8) -> Vec<u8> {
+        let mut full_prefix = Vec::with_capacity(17);
+        full_prefix.push(byte);
+        full_prefix.extend(&self.0);
+        full_prefix
+    }
+}
+
+async fn load_federation_clients(db: &Database) -> Vec<(FederationId, FederationDbPrefix)> {
+    let mut dbtx = db.begin_transaction_nc().await;
+    dbtx.find_by_prefix(&FederationClientPrefix)
+        .await
+        .map(|(k, v)| (k.federation_id, v.db_prefix))
+        .collect::<Vec<_>>()
+        .await
+}
+
+pub fn open_client_db(db: &Database, db_prefix: FederationDbPrefix) -> Database {
+    db.with_prefix(db_prefix.prepend(DbKeyPrefix::ClientDB as u8))
+}
+
+pub async fn try_add_federation_database(
+    db: &Database,
+    federation_id: FederationId,
+    db_prefix: FederationDbPrefix,
+) -> Result<(), FederationDbPrefix> {
+    db.autocommit(
+        |dbtx, _| {
+            Box::pin(async move {
+                if let Some(federation_db_entry) =
+                    dbtx.get_value(&FederationClientKey { federation_id }).await
+                {
+                    return Err(federation_db_entry.db_prefix);
+                }
+
+                dbtx.insert_new_entry(
+                    &FederationClientKey { federation_id },
+                    &FederationClientEntry { db_prefix },
+                )
+                .await;
+
+                Ok(())
+            })
+        },
+        None,
+    )
+    .await
+    .map_err(|e| match e {
+        AutocommitError::CommitFailed { .. } => unreachable!("will keep retrying"),
+        AutocommitError::ClosureError { error, .. } => {
+            // TODO: clean up DB once parallel joins are enabled
+            error
+        }
+    })
+}
+
+pub async fn load_federation_client_databases(db: &Database) -> HashMap<FederationId, Database> {
+    load_federation_clients(db)
+        .await
+        .into_iter()
+        .map(|(federation_id, db_prefix)| (federation_id, open_client_db(db, db_prefix)))
+        .collect()
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+enum DbKeyPrefix {
+    ClientList = 0x00,
+    ClientDB = 0x01,
+    PaymentCodes = 0x02,
+    PaymentCodeNextInvoiceIndex = 0x03,
+    PaymentCodeInvoices = 0x04,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct FederationClientKey {
+    pub federation_id: FederationId,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct FederationClientPrefix;
+
+impl_db_record!(
+    key = FederationClientKey,
+    value = FederationClientEntry,
+    db_prefix = DbKeyPrefix::ClientList,
+);
+impl_db_lookup!(
+    key = FederationClientKey,
+    query_prefix = FederationClientPrefix
+);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct FederationClientEntry {
+    pub db_prefix: FederationDbPrefix,
```

Comment:
What's the purpose of this value? Prefix can be derived from the `FederationId` which is in the key right?

Reply:
The reason why it's random is what you bumped into here: https://github.com/fedimint/fedimint/pull/6855#discussion_r2010334134